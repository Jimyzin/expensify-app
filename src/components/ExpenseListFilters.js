import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setEndDate, setStartDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilter extends React.Component {
    
    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortByChange = (e) => {
        e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();
    };

    render() {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange} 
                />

                <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onSortByChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>

                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filter
});

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);