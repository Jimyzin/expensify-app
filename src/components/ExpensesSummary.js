import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectTotalExpenses from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';

export const ExpensesSummary = ({ totalExpenses, totalCount }) => {
    const total = numeral(totalExpenses / 100).format('$0,0.00');
    const wording = totalCount > 1 ? 'expenses' : 'expense';
    return (
        <div>
            <h1>Viewing {totalCount} {wording} totalling {total}</h1>
        </div>
    );
};

const mapStateToProps = (state) => {
    const selectedExpenses = selectExpenses(state.expenses, state.filter);

    return {
        totalExpenses: selectTotalExpenses(selectedExpenses),
        totalCount: selectedExpenses.length
    };

};

export default connect(mapStateToProps)(ExpensesSummary);