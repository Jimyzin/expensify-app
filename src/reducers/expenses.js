// Expenses Reducer
const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];

        case 'REMOVE_EXPENSE': 
            return state.filter(({ id }) => action.id !== id );

        case 'EDIT_EXPENSE': 
            return state.map((item) => {
                if (action.id === item.id) {
                    return {
                        ...item,
                        ...action.update
                    };
                }
                return item;
            });

        default:
            return state;
    }
};