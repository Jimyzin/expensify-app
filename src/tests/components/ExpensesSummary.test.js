import React from 'react';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';
import { shallow } from 'enzyme';

test('should render multiple expenses', () => {
    const wrapper = shallow( <ExpensesSummary totalExpenses={450000} totalCount={10} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render 1 expense', () => {
    const wrapper = shallow( <ExpensesSummary totalExpenses={450000} totalCount={1} />);
    expect(wrapper).toMatchSnapshot();
});
