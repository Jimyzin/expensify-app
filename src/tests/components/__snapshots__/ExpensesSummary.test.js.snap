// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $4,500.00
  </h1>
</div>
`;

exports[`should render multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    10
     
    expenses
     totalling 
    $4,500.00
  </h1>
</div>
`;
